crypto/schnorr.c

```diff
@@ -0,0 +1,254 @@
+/**
+ * Copyright (c) 2021 The Bitcoin ABC developers
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+ * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "schnorr.h"
+#include "hmac_drbg.h"
+#include "memzero.h"
+#include "rfc6979.h"
+
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+
+int jacobi(const bignum256 *_n, const bignum256 *_k) {
+  int t = 0;
+  int k_is_one;
+  uint32_t r;
+  bignum256 n, k, temp;
+
+  assert(!bn_is_zero(_k) && bn_is_odd(_k));
+
+  bn_copy(_n, &n);
+  bn_copy(_k, &k);
+
+  bn_mod(&n, &k);
+
+  while (!bn_is_zero(&n)) {
+    while (bn_is_even(&n)) {
+      bn_rshift(&n);
+      r = k.val[0] & 0x07;
+      t ^= (r == 3 || r == 5);
+    }
+
+    bn_cmov(&temp, 0, &n, &n);
+    bn_cmov(&n, 0, &k, &k);
+    bn_cmov(&k, 0, &temp, &temp);
+
+    t ^= ((n.val[0] & k.val[0] & 3) == 3);
+
+    bn_mod(&n, &k);
+  }
+
+  k_is_one = bn_is_one(&k);
+
+  // Cleanup
+  memzero(&n, sizeof(n));
+  memzero(&k, sizeof(k));
+  memzero(&temp, sizeof(temp));
+
+  // Map t: [0] => 1, [1] => -1
+  t = -2 * t + 1;
+
+  return k_is_one * t;
+}
+
+int is_non_quad_residue(const bignum256 *n, const bignum256 *prime) {
+  return jacobi(n, prime) == -1;
+}
+
+// Compute sha256(sha256(msg))
+void double_sha256(const uint8_t *msg, uint32_t msg_len,
+                   uint8_t digest[SHA256_DIGEST_LENGTH]) {
+  SHA256_CTX ctx;
+
+  sha256_Init(&ctx);
+  sha256_Update(&ctx, msg, msg_len);
+  sha256_Final(&ctx, digest);
+  sha256_Init(&ctx);
+  sha256_Update(&ctx, digest, SHA256_DIGEST_LENGTH);
+  sha256_Final(&ctx, digest);
+}
+
+/*
+ * Init the HMAC with additional data specific to Schnorr. This prevents from
+ * leaking the private key in the case the same message is signed with both
+ * Schnorr and ECDSA.
+ */
+void init_rfc6979_schnorr(const uint8_t *priv_key, const uint8_t *hash,
+                          rfc6979_state *state) {
+  uint8_t hmac_data[SHA256_DIGEST_LENGTH + 16];
+
+  memcpy(hmac_data, hash, 32);
```

Comment:
`SHA256_DIGEST_LENGTH` is better than 32.

Reply:
Done in bc6267c